subsys/usb/class/bluetooth.c

```diff
@@ -167,32 +169,97 @@ static void hci_rx_thread(void)
 	}
 }
 
+static uint16_t hci_pkt_get_len(struct net_buf *buf,
+				const uint8_t *data, size_t size)
+{
+	uint16_t len = 0;
+	size_t hdr_len = 0;
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_CMD: {
+		struct bt_hci_cmd_hdr *cmd_hdr;
+
+		hdr_len = sizeof(*cmd_hdr);
+		cmd_hdr = (struct bt_hci_cmd_hdr *)data;
+		len = cmd_hdr->param_len + hdr_len;
+		break;
+	}
+	case BT_BUF_ACL_OUT: {
+		struct bt_hci_acl_hdr *acl_hdr;
+
+		hdr_len = sizeof(*acl_hdr);
+		acl_hdr = (struct bt_hci_acl_hdr *)data;
+		len = sys_le16_to_cpu(acl_hdr->len) + hdr_len;
+		break;
+	}
+	case BT_BUF_ISO_OUT: {
+		struct bt_hci_iso_data_hdr *iso_hdr;
+
+		hdr_len = sizeof(*iso_hdr);
+		iso_hdr = (struct bt_hci_iso_data_hdr *)data;
+		len = sys_le16_to_cpu(iso_hdr->slen) + hdr_len;
+		break;
+	}
+	default:
+		LOG_ERR("Unknown bt buffer type");
+		return 0;
+	}
+
+	return (size < hdr_len) ? 0 : len;
+}
+
 static void acl_read_cb(uint8_t ep, int size, void *priv)
 {
-	static uint8_t data[USB_MAX_FS_BULK_MPS];
+	static struct net_buf *buf;
+	static uint16_t pkt_len;
+	uint8_t *data = ep_out_buf;
 
-	if (size > 0) {
-		struct net_buf *buf;
+	if (size == 0) {
+		goto restart_out_transfer;
+	}
 
+	if (buf == NULL) {
+		/*
+		 * Obtain the first chunk and determine the length
+		 * of the HCI packet.
+		 */
 		if (IS_ENABLED(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4) &&
 		    bt_hci_raw_get_mode() == BT_HCI_RAW_MODE_H4) {
 			buf = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, data, size);
+			pkt_len = hci_pkt_get_len(buf, &data[1], size - 1);
```

Comment:
For me it looks like `BT_BUF_H4` buffer type is not handled in `hci_pkt_get_len()` and `pkt_len` will always be 0. Also, where comes the `1` offset here? Could we use `sizeof()`?

Reply:
Refer to: https://github.com/zephyrproject-rtos/zephyr/blob/d85d40c58e862870f74f2c9ad47c2135e94cf80d/subsys/bluetooth/host/hci_raw.c#L139

first byte in H4 mode stores the type.

Reply:
>Also, where comes the 1 offset here?

In this case (BT_HCI_RAW_MODE_H4) it is "like UART transmission" packet and at the index 0 is the packet indicator, which is handled in [bt_buf_get_tx()](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_raw.c#L116).

>For me it looks like BT_BUF_H4 buffer type is not handled in hci_pkt_get_len() and pkt_len will always be 0.

bt_buf_get_tx() also redetermines bt_buf_type. This information is then used in hci_pkt_get_len() to get buffer type.

Reply:
Ok, makes sense now that I looked `hci_raw.c`.